libraries/AP_MSP/AP_MSP.h

```diff
@@ -0,0 +1,144 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <AP_OSD/AP_OSD.h>
+#include "AP_MSP_Telem_Backend.h"
+
+#if HAL_MSP_ENABLED
+
+#define MSP_MAX_INSTANCES 3
+#define ARMING_DISABLED_ARM_SWITCH 1 << 24
+#define ARMING_DISABLE_FLAGS_COUNT 25
+
+#define MSP_OSD_START 2048
+#define MSP_OSD_STEP_X 1
+#define MSP_OSD_STEP_Y 32
+#define MSP_OSD_POS(osd_setting) (MSP_OSD_START + osd_setting->xpos*MSP_OSD_STEP_X + osd_setting->ypos*MSP_OSD_STEP_Y)
+
+using namespace MSP;
+
+class AP_MSP
+{
+public:
+    AP_MSP();
+
+    ~AP_MSP();
+
+    /* Do not allow copies */
+    AP_MSP(const AP_MSP &other) = delete;
+    AP_MSP &operator=(const AP_MSP&) = delete;
+
+    // User settable parameters
+    static const struct AP_Param::GroupInfo var_info[];
+
+    // init - perform required initialisation
+    bool init();
+    void msp_telemetry_callback(uint16_t cmd_msp, uint8_t protocol_instance);
+
+    static AP_MSP *get_singleton(void)
+    {
+        return singleton;
+    }
+
+private:
+    // to access the telemetry_info member variable
+    friend class AP_MSP_Telem;
+    friend class AP_MSP_Telem_DJI;
+    friend class AP_MSP_Telem_Backend;
+
+    enum : uint8_t {
+        BACKEND_MSP,
+        BACKEND_DJI,
+        BACKEND_COUNT
+    };
+
+    enum {
+        OPTION_TELEMETRY_MODE = 1U<<0,
+        OPTION_INVERTED_WIND = 1U<<1
+    };
+
+    AP_Int32 options;
```

Comment:
Would be good to have a comment for each public setting. Just makes it easier to read.
Also new code - I think the convention for new files is member variables starting with"_"?

Reply:
Are you going to have 32 options? I think often these bitsets are smaller

Reply:
not really, 8 bits should be enough

Reply:
private members should be prefixed with _... thanks